refactor: update TLS config validation to use typed matchers for SANs

MayorFaj opened the pull request:
This pull request updates the handling of Subject Alternative Name (SAN) matching in the TLS configuration logic within the `internal/kgateway/extensions2/plugins/backendconfigpolicy/tls.go` file. The changes replace the previous approach of using string matchers with a new typed SAN matcher implementation to improve specificity and correctness in SAN validation.

### Updates to SAN Matching Logic:

* **Refactor SAN Matching Method**: The function `verifySanListToMatchSanList` was replaced by `verifySanListToTypedMatchSanList` to use `envoyauth.SubjectAltNameMatcher` instead of `envoymatcher.StringMatcher`. This introduces typed SAN matching, specifying `SubjectAltNameMatcher_DNS` as the SAN type and maintaining exact string matching for SAN validation.

* **Integration of Typed SAN Matchers**: Updated references in `translateTLSConfig` to use the newly implemented `verifySanListToTypedMatchSanList`. The `MatchSubjectAltNames` field was replaced with `MatchTypedSubjectAltNames` in the TLS validation context to align with the new typed SAN matcher logic.



# Change Type

```
/kind deprecation
```

# Changelog

```release-note
Fixes deprecation warning by replacing deprecated MatchSubjectAltNames field with MatchTypedSubjectAltNames in TLS validation context for BackendConfigPolicy.
```

closes #11477

timflannagan reviewed (commented):
LGTM - I'm surprised we don't see any of the golden files being changed here though. cc @puertomontt. Maybe we have a gap in test coverage?

Additionally, I think we can use the `/kind fix` type here instead of the deprecation kind too.

timflannagan commented:
@MayorFaj WYDT? I think we either track the need for this type of test coverage or add test coverage to this branch to validate the changes work as expected. I'm cool with either approach, but don't want it to slip under the radar.

MayorFaj commented:
> @MayorFaj WYDT? I think we either track the need for this type of test coverage or add test coverage to this branch to validate the changes work as expected. I'm cool with either approach, but don't want it to slip under the radar.

I am currently working on adding the test to this branch

timflannagan reviewed (commented):
Sorry for the slow reply - thanks for adding that unit test. I should've clarified that when I meant adding a test case, I meant adding one in the gateway translator suite, which allows us to see the output xDS configuration produced from gateway translation. Is it possible to add a simple test case to that suite as well?

MayorFaj commented:
> Sorry for the slow reply - thanks for adding that unit test. I should've clarified that when I meant adding a test case, I meant adding one in the gateway translator suite, which allows us to see the output xDS configuration produced from gateway translation. Is it possible to add a simple test case to that suite as well?

updated, and I have added docs on steps to achieve that, I figured that part was missing. Would be useful for ref.

timflannagan commented on internal/kgateway/translator/gateway/testutils/inputs/backendconfigpolicy/tls-san.yaml:
Thanks for adding this test case. This is great. I'm looking at the tls-sans-out.yaml file and I'm seeing the SANs related changes surfaced in that file. Any idea what's happening here?

timflannagan replied:
Ah maybe it's due to invalid Secret data?

```
conditions:
    - lastTransitionTime: "2025-07-02T12:23:47Z"
      message: 'tls: failed to find any PEM data in certificate input'
      observedGeneration: 1
      reason: Invalid
      status: "False"
      type: Accepted
    controllerName: kgateway.dev/kgateway
```

That's what I'm seeing when I deploy these changes to a local cluster.

MayorFaj replied:
Thanks for this.I have updated the yaml with the cert, and regenerated. The earlier issue seems to be with format or encoding.
I will move the docs to a new PR as well

timflannagan reviewed (approved):
Thanks! Confirmed locally that the fix works:

```diff
+++ b/internal/kgateway/translator/gateway/testutils/outputs/backendconfigpolicy/tls-san.yaml
@@ -66,13 +66,9 @@ Clusters:
               3teRYW/cEDQ1PTgLfAkb7Fk=
               -----END PRIVATE KEY-----
         validationContext:
-          matchTypedSubjectAltNames:
-          - matcher:
-              exact: test.example.com
-            sanType: DNS
-          - matcher:
-              exact: api.example.com
-            sanType: DNS
+          matchSubjectAltNames:
+          - exact: test.example.com
+          - exact: api.example.com
```
